internal_docs/protocol.md

```diff
@@ -0,0 +1,60 @@
+# RTC Engine Signaling Protocol
+
+### Joining the server
+
+```mermaid
+sequenceDiagram
+    client->>server: join
+    server->>client: peerAccepted
+    server->>other_client: peerJoined
+```
+
+### Leaving the server
+
+```mermaid
+sequenceDiagram
+    client->>server: leave
+    server->>other_client: peerLeft
+    rect rgb(135, 204, 232)
+        note right of server: renegotiation
+        server->>other_client: custom(offerData)
+        other_client->>server: custom(sdpOffer)
+        par
+            other_client->>server: custom(candidate)
+        and
+            server->>other_client: custom(sdpAnswer)
+            server->>other_client: custom(candidate)
+        end
+    end
+```
+
+### Adding a track
+
+```mermaid
+sequenceDiagram
+    client->>server: join
+    server->>client: peerAccepted
+    server->>other_client: peerJoined
+    client->>server: renegotiateTracks
+    server->>client: custom(offerData)
+    client->>server: custom(sdpOffer)
+    par
+        client->>server: custom(candidate)
+    and
+        server->>client: custom(sdpAnswer)
+        server->>client: custom(candidate)
+        server->>other_client: tracksAdded
+        server->>client: tracksAdded 
+        rect rgb(135, 204, 232)
+            note right of server: renegotiation
```

Comment:
Should this be nested inside the parallel block?

Reply:
Hard to say, some blocks are indented but some are not. E.g. 

![image](https://user-images.githubusercontent.com/24651552/170988363-42c09d71-f771-4c26-8bf5-a1dd3a6376f8.png)

or 

![image](https://user-images.githubusercontent.com/24651552/170988452-10e241e2-f2b8-4ecc-b1fb-23e17dcb0a35.png)


But 

![image](https://user-images.githubusercontent.com/24651552/170988413-1a9f5cb5-47b6-41bc-ba9e-dd9c0f75a174.png)

I would keep the indentation

Reply:
I haven't meant the code formatting (it seems fine) but rather the concept - should the whole renegotiation block be put inside `par` block? I mean I think the renegotiation with `other_client` cannot happen in parallel to sending candidates from `client` - the send of `trackAdded` to `other_client` should end that parallel block - am I correct?

Reply:
Ach, so it happens in parallel. We start restarting other participants connections as fast as possible i.e. after generating SDP answer for `client`